test/unit/key_def.c

```diff
@@ -0,0 +1,267 @@
+	const char *parts = mp_buf;
+	uint32_t part_count = mp_decode_array(&parts);
+	struct key_part_def *part_def = region_alloc(
+		region, sizeof(*part_def) * part_count);
+	fail_if(part_def == NULL);
+	fail_if(key_def_decode_parts(part_def, part_count, &parts,
+				     /*fields=*/NULL, /*field_count=*/0,
+				     region) != 0);
+
+	/* Create a key def. */
+	struct key_def *def = key_def_new(part_def, part_count, for_func_index);
+	fail_if(def == NULL);
+	key_def_update_optionality(def, 0);
+
+	region_truncate(region, region_svp);
+	return def;
+}
+
+/** Creates a key_def from a MsgPack format (see mp_format). */
+static struct key_def *
+test_key_def_new(const char *format, ...)
+{
+	va_list ap;
+	va_start(ap, format);
+	struct key_def *def = test_key_def_new_va(format, ap,
+						  /*for_func_index=*/false);
+	va_end(ap);
+	return def;
+}
+
+/** Creates a functional index key_def from a MsgPack format (see mp_format). */
+static struct key_def *
+test_key_def_new_func(const char *format, ...)
+{
+	va_list ap;
+	va_start(ap, format);
+	struct key_def *def = test_key_def_new_va(format, ap,
+						  /*for_func_index=*/true);
+	va_end(ap);
+	return def;
+}
+
+/**
+ * Checks that tuple_compare_with_key with cmp_def of functional index
+ * returns the same result as comparison of concatenated func and primary keys.
+ */
+static void
+test_check_tuple_compare_with_key_func(
+	struct key_def *cmp_def, struct tuple *tuple,
+	char *func_key, struct key_def *model_def,
+	struct tuple *model, const char *key)
+{
+	fail_unless(cmp_def->for_func_index);
+	fail_if(model_def->for_func_index);
+	const char *key_parts = key;
+	uint32_t part_count = mp_decode_array(&key_parts);
+	int a = tuple_compare_with_key(tuple, (hint_t)func_key, key_parts,
+				       part_count, HINT_NONE, cmp_def);
+	int b = tuple_compare_with_key(model, HINT_NONE, key_parts,
+				       part_count, HINT_NONE, model_def);
+	a = a > 0 ? 1 : a < 0 ? -1 : 0;
+	b = b > 0 ? 1 : b < 0 ? -1 : 0;
+	is(a, b, "tuple_compare_with_key_func(%s/%s, %s) = %d, expected %d",
+	   tuple_str(tuple), mp_str(func_key), mp_str(key), a, b);
+}
+
+static void
+test_func_compare_with_key(void)
+{
+	plan(10);
+	header();
+
+	struct key_def *def = test_key_def_new_func(
+		"[{%s%u%s%s}{%s%u%s%s}]",
+		"field", 0, "type", "unsigned",
+		"field", 1, "type", "string");
+	/* Skip first field to check if func comparator can handle this. */
+	struct key_def *pk_def = test_key_def_new(
+		"[{%s%u%s%s}{%s%u%s%s}]",
+		"field", 1, "type", "unsigned",
+		"field", 2, "type", "string");
+	struct key_def *cmp_def = key_def_merge(def, pk_def);
+	/*
+	 * Model def is a copy of cmp_def, but not for_func_index, and hence
+	 * it has general implementation of tuple_compare_with_key method.
+	 */
+	struct key_def *model_def = test_key_def_new(
+		"[{%s%u%s%s}{%s%u%s%s}{%s%u%s%s}{%s%u%s%s}]",
+		"field", 0, "type", "unsigned",
+		"field", 1, "type", "string",
+		"field", 3, "type", "unsigned",
+		"field", 4, "type", "string");
```

Comment:
The test still contains some logic of primary key comparison check. I think it's better not to export it at all.

Reply:
Agreed. Replaced the test with comparator tests from #8915, they fully cover the `func_index_compare_with_key` and other comparators.

Reply:
Wow, it's quite a large test suite. Let's just export my commit without tests :)

Reply:
The test suit will be reviewed in the scope of #8915, and the commit is trivial - just factoring out some logic. And it will be covered by other tests (`key_compare_parts` wrapper is already covered, `key_compare_and_skip_parts` will be covered in the next commit).